### PR TITLE
Adds CredentialRetrieverFactory with initial one for Docker credential helpers.

### PR DESCRIPTION
**Pull request by coollog**
Part of #774 

This also changes many classes to take the path to a credential helper executable rather than just a credential helper suffix. The standard `docker-credential-` prefix is encapsulated in `DockerCredentialHelperFactory` now.

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/RetrieveRegistryCredentialsStep.java [hunk 1]
```diff
@@ -196,10 +198,12 @@ Authorization retrieveFromCredentialHelper(String credentialHelperSuffix)
     buildLogger.info("Checking credentials from docker-credential-" + credentialHelperSuffix);
 
     try {
-      Authorization authorization =
+      Credentials credentials =
           dockerCredentialHelperFactory
               .newDockerCredentialHelper(registry, credentialHelperSuffix)
               .retrieve();
+      Authorization authorization =
+          Authorizations.withBasicCredentials(credentials.getUsername(), credentials.getPassword());
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/RetrieveRegistryCredentialsStep.java]: Wouldn't `withBasicCredentials(credentials)` would be simpler and match the signature?

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java [hunk 2]
```diff
@@ -27,7 +27,9 @@
    *
    * @return the fetched credentials or {@code null} if no credentials could be fetched with this
    *     provider
+   * @throws Exception if the execution should fail instead of trying other credential retrieval
+   *     methods
    */
   @Nullable
-  Credentials retrieve();
+  Credentials retrieve() throws Exception;
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: I think you should introduce a specific exception.

File: jib-core/src/main/java/com/google/cloud/tools/jib/http/Authorization.java [hunk 3]
```diff
@@ -46,4 +48,21 @@ public String getToken() {
   public String toString() {
     return scheme + " " + token;
   }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/http/Authorization.java]: For consistency, `other instanceof Authorization`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelperFactory.java [hunk 4]
```diff
@@ -16,18 +16,29 @@
...
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
 /** Factory class for constructing {@link DockerCredentialHelper}. */
 public class DockerCredentialHelperFactory {
 
+  public static final String CREDENTIAL_HELPER_PREFIX = "docker-credential-";
+
   public DockerCredentialHelperFactory() {}
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelperFactory.java]: I'm not sure I see the value in having a separate factory — when there's nothing to configure on the factory — vs static methods on `DockerCredentialHelper`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentDockerCredentialHelperException.java [hunk 5]
```diff
@@ -16,10 +16,12 @@
...
 package com.google.cloud.tools.jib.registry.credentials;
 
+import java.nio.file.Path;
+
 /** Thrown because the requested credential helper CLI does not exist. */
 public class NonexistentDockerCredentialHelperException extends Exception {
 
-  NonexistentDockerCredentialHelperException(String credentialHelperSuffix, Throwable cause) {
-    super("The system does not have docker-credential-" + credentialHelperSuffix + " CLI", cause);
+  NonexistentDockerCredentialHelperException(Path credentialHelper, Throwable cause) {
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentDockerCredentialHelperException.java]: `CannotFindDockerCredentialHelperException`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentServerUrlDockerCredentialHelperException.java [hunk 6]
```diff
@@ -16,11 +16,13 @@
 
 package com.google.cloud.tools.jib.registry.credentials;
 
+import java.nio.file.Path;
+
 /** Thrown because the credential helper does not have credentials for the specified server URL. */
 public class NonexistentServerUrlDockerCredentialHelperException extends Exception {
 
   NonexistentServerUrlDockerCredentialHelperException(
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentServerUrlDockerCredentialHelperException.java]: This name is also a bit hard to parse.  `UnknownServerUrlException`?  Maybe the docker stuff should be in a separate package?

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java [hunk 7]
```diff
@@ -27,7 +27,9 @@
    *
    * @return the fetched credentials or {@code null} if no credentials could be fetched with this
    *     provider
+   * @throws Exception if the execution should fail instead of trying other credential retrieval
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: Can't get what this really means. Does it mean

"the retrieval attempt this `CredentialRetriever` tried failed if you got an `Exception` (but then what "other credential methods" mean? Is `CredentialRetriever` supposed to try multiple methods in general? Even so, I guess it's an implementation detail.)

or

"if you as a caller is in the middle of trying multiple `CredentialRetriever`s and got any `Exception` from any of them, you as a caller should stop trying other retrievers and "fail" right away."

In any case, I think the consumer of the Javadoc of an interface should be the caller in general.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+   * @return a new {@link CredentialRetriever}
+   * @see <a
+   *     href="https://github.com/docker/docker-credential-helpers#development">https://github.com/docker/docker-credential-helpers#development</a>
+   */
+  public CredentialRetriever dockerCredentialHelper(Path credentialHelper) {
+    return dockerCredentialHelper(credentialHelper, new DockerCredentialHelperFactory());
+  }
+
+  @VisibleForTesting
+  CredentialRetriever dockerCredentialHelper(
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: Do I have this straight: this is a factory that creates a retriever to defer a retrieval using a factory to create a helper?  It feels like there is a bit too many levels of indirection here.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java [hunk 9]
```diff
@@ -54,30 +54,30 @@
...
    * echo -n <server URL> | docker-credential-<credential helper suffix> get
    * }</pre>
    *
+   * @return the Docker credentials by calling the corresponding CLI
    * @throws IOException if writing/reading process input/output fails.
    * @throws NonexistentServerUrlDockerCredentialHelperException if credentials are not found.
    * @throws NonexistentDockerCredentialHelperException if the credential helper CLI doesn't exist.
    */
-  public Authorization retrieve()
+  public Credentials retrieve()
```
Comment (briandealwis) [inline jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java]: Can't DockerCredentialHelper be a `CredentialsRetriever`?

**Comment by coollog** [comment]
@chanseokoh Yea, I'm thinking maybe we could return an error type in the return `retrieve` (aka in `CredentialRetriever`), but that since `CredentialRetriever` is intended to be implemented by users as well, it might make implementations more complicated than they need to be.

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java (hunk 7 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: Hmm, I was intending the consumer of this javadoc to be the user implementing the interface. The intention was:

"If you decide to have this throw an exception, it indicates that something went awry and execution should not continue (whereas if you decide to have this return null, it just indicates that this retrieval failed to retrieve any credentials, but execution should continue on)."

File: jib-core/src/test/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    Mockito.when(
+            mockDockerCredentialHelperFactory.newDockerCredentialHelper(
+                "registry", Paths.get("docker-credential-helper")))
+        .thenReturn(mockDockerCredentialHelper);
+    Mockito.when(mockDockerCredentialHelper.retrieve()).thenReturn(expectedCredentials);
+
+    Assert.assertEquals(
+        expectedCredentials,
+        credentialProviderFactory
```
Comment (briandealwis) [inline jib-core/src/test/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactoryTest.java]: This seems the opposite of what I'd expect: I thought you'd create a DockerCredentialHelper and then ask it to get credentials for the various image references.

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java (hunk 7 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: Reworded it a bit.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: I believe Javadoc of almost every class or interface is intended for the users of them. If you want to direct implementors to something specific or certain implementation details, I'd say, e.g., "implementation of this class should be thread-safe", "note to implementors:", or such.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/RetrieveRegistryCredentialsStep.java (hunk 1 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/RetrieveRegistryCredentialsStep.java]: Currently, we have the `Credentials` in the `configuration` package, whereas `Authorizations` is in the `http` package. I'm not sure if we want `Authorizations` to have a dependency on `Credentials` yet.

File: jib-core/src/test/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactoryTest.java (hunk 10 above)
Comment (chanseokoh) [reply jib-core/src/test/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactoryTest.java]: Hmm... interesting point. Honestly, I'm overall not familiar how all the credential-related classes are interfaced.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java (hunk 8 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java]: Yea, filed #830 to see about removing the `DockerCredentialHelperFactory`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java (hunk 9 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java]: Yep, we can definitely make it an implementation of `CredentialsRetriever`. I added it as a task to do after we remove the uses in `RetrieveRegistryCredentialsStep`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentDockerCredentialHelperException.java (hunk 5 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentDockerCredentialHelperException.java]: Filed as separate issue: #832

File: jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentServerUrlDockerCredentialHelperException.java (hunk 6 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/NonexistentServerUrlDockerCredentialHelperException.java]: ditto #832

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java (hunk 2 above)
Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: Something like a `CredentialRetrievalException`?

**Comment by coollog** [comment]
Requesting another approval since I made a few changes.

File: jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java [hunk 11]
```diff
@@ -25,9 +25,14 @@
...
+   * {@link CredentialRetriever} (and so other credential retrieval methods may be tried), or throw
+   * an exception something went wrong when fetching the credentials.
+   *
    * @return the fetched credentials or {@code null} if no credentials could be fetched with this
    *     provider
+   * @throws Exception if the credential retrieval encountered an exception
    */
   @Nullable
-  Credentials retrieve();
+  Credential retrieve() throws Exception;
```
Comment (TadCordle) [inline jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: I agree with Brian's comment about using a more specific exception.

Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/configuration/credentials/CredentialRetriever.java]: Added that to the TODOs of #832
